sdk/python/lib/pulumi/runtime/resource.py

```diff
@@ -11,152 +11,141 @@
-        # Neither pylint nor I are the only ones who find this confusing:
-        # https://github.com/grpc/grpc/issues/10885#issuecomment-302581315
-        # pylint: disable=no-member
-        if exn.code() == grpc.StatusCode.UNAVAILABLE:
-            sys.exit(0)
-
-        # If the RPC otherwise failed, re-throw an exception with the message details - the contents
-        # are suitable for user presentation.
-        raise Exception(exn.details())
-
-
-    # If the invoke failed, raise an error.
-    if resp.failures:
-        raise Exception('invoke of %s failed: %s (%s)' % (tok, resp.failures[0].reason, resp.failures[0].property))
-
-    # Otherwise, return the output properties.
-    retobj = getattr(resp, 'return')
-    if retobj:
-        return rpc.deserialize_resource_props(retobj)
-
-    return {}
-
-class RegisterResourceResult(object):
-    """
-    RegisterResourceResult contains the assigned URN, the ID -- if applicable -- and the resulting resource
-    output properties, representing a resource's state after registration has completed.
-    """
-    def __init__(self, urn, id, outputs):
-        self.urn = urn
-        self.id = id
-        self.outputs = outputs
-
-def register_resource(typ, name, custom, props, opts):
+import sys
+import traceback
+from typing import Optional, Awaitable, Any, Callable, List, NamedTuple, Dict, Set, Union, TYPE_CHECKING
+
+from . import rpc, settings, known_types
+from .. import log
+from ..runtime.proto import resource_pb2
+from .rpc_manager import RPC_MANAGER
+
+if TYPE_CHECKING:
+    from .. import Resource, ResourceOptions
+    from ..output import Output, Inputs
+
+
+class ResourceResolverOperations(NamedTuple): #TODO(sean) rename
+    parent_urn: Optional[str]
+    serialized_props: Dict[str, Any]
+    dependencies: Set[str]
+
+
+# Prepares for an RPC that will manufacture a resource, and hence deals with input and output properties.
+async def prepare_resource(ty: str, props: 'Inputs', opts: Optional['ResourceOptions']) -> ResourceResolverOperations:
+    log.debug(f"resource {props} preparing to wait for dependencies")
+    # Before we can proceed, all our dependencies must be finished.
+    explicit_urn_dependencies = []
+    if opts is not None and opts.depends_on is not None:
+        dependent_urns = list(map(lambda r: r.urn.future(), opts.depends_on))
+        explicit_urn_dependencies = await asyncio.gather(*dependent_urns)
+
+    # Serialize out all our props to their final values.  In doing so, we'll also collect all
+    # the Resources pointed to by any Dependency objects we encounter, adding them to 'implicit_dependencies'.
+    implicit_dependencies: List[Resource] = []
+    serialized_props = await rpc.serialize_properties(props, implicit_dependencies)
+
+    # Wait for our parent to resolve
+    parent_urn = ""
+    if opts is not None and opts.parent is not None:
+        parent_urn = await opts.parent.urn.future()
+    elif ty != "pulumi:pulumi:Stack": # TODO(sean)
```

Comment:
Is this "TODO" intentional?

Reply:
yes, I'm planning on circling back and removing or commenting my todos as I go. I left it in here so I don't forget to revisit.

I'm not sure how we avoid the Stack resource from parenting to itself in the Node SDK. I found that the "ty" check was necessary here to prevent that and I'm not sure why we don't need it in Node.